### PR TITLE
Merge constructor augmentation changes into the augmentations proposal

### PR DESCRIPTION
**Pull request by jakemac53**
Merges https://github.com/dart-lang/language/issues/4063 into the augmentations feature spec, with modifications from the comments. 

- Changes how non-redirecting generative constructor augmentations work, roughly:
  - Remove special meaning of `augmented` in these constructor bodies.
  - Remove requirement that initializing formal and super parameters are a part of the "identity" of a constructor call, so future augmentations can change them.
    - Instead, only a single declaration opts a parameter into being either an initializing formal or super parameter.
    - All super parameters must appear in the same declaration, but it need not be the base one.
    - Any declaration introducing super parameters must come before any super constructor invocations. 
  - Constructor bodies are invoked in order, starting at the base declaration and then going through the augmentations in first to last order.
- Describes in detail the non-redirecting generative constructor invocation semantics.

I left off the requirement that the super constructor invocation must appear in the same declaration as any super parameters, and instead allow it to come in any subsequent declaration. It was simply easier to specify, and seemed fairly harmless, but let me know if you think this should change.

cc @lrhn should I bring in the "details of the definitions" section as well or is this sufficient?

### CODE REVIEW

File: working/augmentation-libraries/feature-specification.md [hunk 1]
```diff
@@ -932,8 +1082,13 @@ potentially non-redirecting property of the constructor.
 
 It is a compile-time error if:
 
-*   The augmented constructor has an initializer list or a body, or it has a
-    redirection.
+*   The augmented constructor has any initializers (including a redirecting
+    initializer).
```
Comment (eernstg) [inline working/augmentation-libraries/feature-specification.md:1086]: I don't think there is such a thing as a redirecting initializer. I think it could just be deleted:
```suggestion
*   The augmented constructor has any initializers.
```

File: working/augmentation-libraries/feature-specification.md [hunk 2]
```diff
@@ -932,8 +1082,13 @@ potentially non-redirecting property of the constructor.
...
+*   The augmented constructor has a body.
+*   The augmented constructor has a redirection.
+
+The new redirecting generative constructor now behaves exactly like any other
```
Comment (eernstg) [inline working/augmentation-libraries/feature-specification.md:1090]: Perhaps:
```suggestion
This redirecting generative constructor now behaves exactly like any other
```

**Review by eernstg — COMMENTED** [review]
Added a bunch of comments. ;-)

File: working/augmentation-libraries/feature-specification.md [hunk 3]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+                *v*.
 
-    *   Local variables in scope where `augmented()` is evaluated are not in
-        scope for the execution of the augmented constructor's body.
+        *   Bind the name of the parameter to *v* in the initializer-list scope.
 
-*   Add initializers to the initializer list. If the augmenting constructor has
-    an initializer list then:
+    *   Otherwise the parameter is a normal parameter. Bind the name of the
+        parameter to *v* in the parameter scope.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:988]: No, we are explicitly allowing this now (it wasn't allowed before).

It allows a macro or augmentation to add an `extends` clause and simultaneously add the super call, while taking advantage of the super parameter syntax to construct that call by augmenting some parameters to be super parameters.

File: working/augmentation-libraries/feature-specification.md [hunk 4]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+        value *m*, if there is a second operand, otherwise let *m* be the
+        `null` value. Throw an `AssertionError` with *m* as message.
+
+    *   If variable intializer entry, evaluate expression in initializer list
+        scope to value *v*. Initialize the variable of *o* corresponding to
+        variable with the initializer entry name in *C* to the value *v*.
+
+*   If *d* has a *superConstructorInvocation*.
+
+    * It is an error if *s* is already initialized.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:1003]: Yes, if multiple augmentations have super initializes. This is just making that an error.

File: working/augmentation-libraries/feature-specification.md [hunk 5]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+
+    * Initialize *s* to the *superConstructorInvocation* of *d*.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md]: I am open to suggestions, but I didn't really see a good way otherwise...

File: working/augmentation-libraries/feature-specification.md [hunk 6]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+
+*   If *d* has an *augmentedDefinition* *d’*
+
+    *   Invoke *d’* on *C\<T1,…,Tn>* with arguments *L*, super-arguments *S*,
+        and super constructor invocation *s* to initialize *o*.
+
+    *   *This recurses on the augmented definition*.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:1012]: Well, a static error isn't called out anywhere right now I don't believe.

It certainly could be a static error though. Maybe I just mention on this line that this error should be reported at compile time?

File: working/augmentation-libraries/feature-specification.md [hunk 7]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+
+*   Otherwise:
+
+    *   Let *U* be be the instantiated *superclass definition* of
+        *C*\<*T1*,…,*Tn*\>.
+
+    *   This may be a synthetic definition (for an anonymous mixin application)
+        computed from the declared superclass and declared mixins of *C*. _(Or
+        we can do the short-circuiting here since mixin applications can only)
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md]: Yeah I don't think this is really needed, removed.

File: working/augmentation-libraries/feature-specification.md [hunk 8]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+
+    *   If *s* is initialized:
+
+        *   Evaluate each argument of the argument list of the super-constructor
+            invocation in the initializer list scope, in source order, and set
+            the entry of *S* with the same position or name to the resulting
+            value.
+
+        *   Let *g* be the name of superclass constructor of *U* targeted by the
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md]: done

File: working/augmentation-libraries/feature-specification.md [hunk 9]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+
+*   For each *instance variable definition* of *C* in *base declaration order*
+    (the order of the *base* declaration in the total traversal ordering of a
+    library), where the variable definition *has an initializer expression*, and
+    is not declared `late`:
+
+    *   Evaluate the initializer expression of that definition to a value *v*
+        (taking any augmentations of it into account).
+
+    *   Initialize the corresponding instance variable of *o* to the value *v*.
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply working/augmentation-libraries/feature-specification.md:914]: I guess we need to define "class definition", such that it includes the inherited variable definitions?

File: working/augmentation-libraries/feature-specification.md [hunk 10]
```diff
@@ -869,49 +879,189 @@ It is a compile-time error if:
...
+
+*   Then invocation of *d* completes normally.
+
+The consequence of this definition is an execution order for initialization of
+an instance of a class of:
+
+*   Field initializers of class, from all augmentations, in “base declaration
+    source order”.
+
+*   Augmenting declaration parameter lists and initializer lists, for
```
[the remark this answers is not in the document]
Comment (lrhn) [reply working/augmentation-libraries/feature-specification.md:1048]: The "evaluate parameter list" here is actually performing "binding actuals to formals", which is the step that handles initializing formals and super parameters.

Or in more detail... When you execute a (non-redirecting generative) contructor of a class to initialize the object, given an argument list, it does the following:
- For every instance variable of the class, in source order of their base declarations,
  execute any initializer expressions for the variable.
  - Start with the last variable declaration for that variable that has an initializing expression.
  - If that declaration is in an augmenting declaration, and it contains `augmented`,
    then each occurrence of `augmented` evaluates the initializer expression of the
    next prior declaration of that variable that has an initializing expression.
  - Then initialize the variable to the resulting value.
- Then start with the last (in source order) declaration of that constructor, and execute that
  to initialize the new object with the given argument list.
- Executing a constructor declaration in this way means:
    - Bind the actual argument list to the declared formal parameters of that constructor 
      declaration. This handles initializing formals and super parameters (which assign the 
      value to some position in a super constructor invocation argument list), and binds
      names to values in the parameter and initializer list scopes.
    - Execute the initializer list of that constructor declaration in the intializer list scope.
    - If the constructor declares a super-constructor invocation, remember that superclass
      constructor, and the accumulated super-constructor argument list.
    - If the declaration is an augmenting constructor declaration, then
      execute (recursively) the augmented declaration to initialize the same object
      with the same argument list.
     - Otherwise execute the remembered superclass constructor with the constructed
       argument list to initialize the same object. (If there is no remembered superclass
       constructor, execute the unnamed constructor of the superclass with the collected
       argument list).
    - When either of these complete, execute the body (if any) of this declaration in the 
      parameter scope of.
    - Then the execution of this constructor declaration completes.
- When the last constructor declaration execution has completed, the execution of the
  class constructor completes.

**Comment by eernstg** [comment]
(I'm OOO for a couple of days, I'll return to this on Monday.)

**Review by eernstg — COMMENTED** [review]
I think this is a heroic effort to specify the semantics of generative non-redirecting constructor invocations based on the syntax as specified by the programmer.

I think it almost works.

At the same time, it illustrates that this approach is quite costly in terms of detailed elements that are specified from scratch. It would be very helpful to introduce a number of abstractions (search for 'declarative specification' to see one example).

One of the consequences of this is that we get the surprising evaluation order for initializer list elements (search for 'i7' in the comments to see more). I certainly think we'd deliver a better quality of service if we preserve the evaluation order for a constructor which is declared in one piece.

Another thing I'm worried about is the fact that initializing formals and super-parameters will block further augmentation. I would certainly expect that an initializing formal could be introduced (like `A(this.x);`) and later augmentations could just confirm its status (like `@metadata augment A(this.x);`), but this does not seem to be possible.

In other words, every parameter of the form `this.x` or `super.x` _must occur in the laste augmentation_. Is it really scalable (and compatible with macros) to have this requirement?

**Comment by jakemac53** [comment]
I am closing this PR, a [simplified version](https://github.com/dart-lang/language/pull/4192) has been merged which doesn't describe the full process of running a constructor, and just contains the key changes surrounding the use of `augmented` in constructor bodies (not allowed any more for generative, non-redirecting constructors).
